Development_Documents/BOINC_Flow.md

```diff
@@ -0,0 +1,47 @@
+
+## Master Branch
+The master branch is the primary branch in the project.  Committers will never commit to the master branch directly.  Any code merged into master is expected to go through peer review, be unit tested and have had existing automation tests run successfully against the new code before it is merged.  As a result, the master branch is expected to be reliable and is what developers should usually branch from when creating other branches.  
+
+## Feature Branches
+Developers create feature branches when they want to modify the code base.  If they are a committer, then they can use the [shared repository model](https://gist.github.com/seshness/3943237) and create the feature branch within the BOINC project directly.  If they are not a committer then they should follow the [fork and pull approach](https://gist.github.com/Chaser324/ce0505fbed06b947d962) to create a branch and submit a pull request.
+
+Features branches are merged into master via a pull request.  Developers should understand what is expected before submitting a pull request by reviewing the [Development Workflow](Development_Workflow.md) document.  
+
+Once a feature branch has been merged into master it should be deleted.
+
+<p align="center">
+<img src="images/FeatureBranch.png" alt="Diagram of a Feature Branch"/>
+</p>
+
+## Bugfix Branch
+Bugfix branches are basically the same as feature branches except that created to fix a bug rather than implement a new feature.  This difference is mostly important when we talk about release branches later on in this document.
+
+<p align="center">
+<img src="images/BugfixBranch.png" alt="Diagram of a Bugfix Branch"/>
+</p>
+
+## Release Branches
+BOINC has two types of releases: client releases and server releases.  Each of these require additional integration and end user testing that is not usually done for each individual feature or bugfix branches.  This additional testing needs to be conducted against a stable code base (code freeze) while also allowing development for future work to continue.  The way to do this is to create a release branch.  The release branch freezes the code that is being tested.  Other development can continue on the master branch without impacting the release branch.
+
+A release branch is created off of the master branch.  Since the release branch is considered feature frozen, new code should only be added to the branch if the release manager determines that it is necessary.  
+
+Any code that is added to a release branch after it is created should be a cherry-pick of a merge commit that is present on the master branch.  The merge commit would be the result of a merging a bugfix or feature branch that was created and merged as normal.
+
```

Comment:
Cherry-picking a (random) commit from another branch (here: master) might not be the best possible solution. There is no (technically supported) way to ensure that
1. this single commit that works on one branch doesn't break the (build of) the release branch and
2. everything that is needed to fix the problem is contained in the single commit
For creating a bugfix branch I would recommend to fork from the last common commit of the master and (last) release branch, and then separately merge into master and release branch via PRs.

Reply:
So it isn't cherry-picking a random commit but rather cherry-picking a merge commit.  The community has been doing pretty well with making their feature branches atomic so the chance of bringing in misc code is significantly reduced (not eliminated).

If you can provide any easy way (one line command) to identify the commit where a branch was forked, then I would be happy to add diagrams and instructions stating that if you are making fix intended for a release branch then you should create your branch from that commit.

Reply:
`git merge-base master <branch/HEAD>`